# Song Genre Classifier (via Lyric Analysis)

**By Josiah Nielsen**

**For: Collecting and Analyzing Big Data**

This notebook contains code to analyze a dataset of song lyrics and genres. I will attempt to classify the genre of each song using only the song lyrics, through the use of various algorithms and feature extraction.  

First, I will preprocess the data. This consists of removing punctuation and stop words from the lyrics. The lyrics are then tokenized and stemmed.

Then, I performed classification (Naive Bayes, SVM, XGBoost, RandomForests, etc), using Count and TF-IDF Vectorization before classification. 

Finally, I performed feature extraction using Parts of Speech and Embeddding (Word2Vec), followed by classification modelling using these features. These models performed poorly, and only my Gradient Boosting Classifer with Word2Vec has been retained to demonstrate use of other methods.

The XGBoost classifier w/ TF-IDF vectorization performed the best out of all models, with a 63% accuracy. The SVM classifier performed almost as good, with an accuracy of 62%.

In [ ]:
#Import dependencies
import numpy as np
import pandas as pd
pd.options.mode.chained_assignment = None

import nltk
nltk.download('stopwords')
from nltk import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer

import json
import re
import string

import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)
warnings.simplefilter(action='ignore', category=FutureWarning)

from sklearn.multiclass import OneVsRestClassifier
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer 
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

from xgboost import XGBClassifier

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


#Data Preprocessing

In [ ]:
#Load Data
df = pd.read_csv("SongLyrics.csv")

In [ ]:
#Remove songs with too few lyrics
df['word_count'] = df['lyrics'].str.split( ).str.len()
df = df[df['word_count'] > 50]

In [ ]:
df.genre.value_counts()

Country    12339
Pop        12070
Rock       11875
Metal      11346
Hip-Hop    10651
Name: genre, dtype: int64

In [ ]:
#Create copy of unprocessed dataframe
df_unprocessed = df 

In [ ]:
#Define function for preprocessing the lyrics
def preprocessText(text, remove_stops=True):
    
    # Remove everything between hard brackets
    text = re.sub(pattern="\[.+?\]( )?", repl='', string=text)

    # Change "walkin'" to "walking", for example
    text = re.sub(pattern="n\\\' ", repl='ng ', string=text)

    # Remove x4 and (x4), for example
    text = re.sub(pattern="(\()?x\d+(\))?", repl=' ', string=text)

    # Fix apostrophe issues
    text = re.sub(pattern="\\x91", repl="'", string=text)
    text = re.sub(pattern="\\x92", repl="'", string=text)
    text = re.sub(pattern="<u\+0092>", repl="'", string=text)
    
    # Make lowercase
    text = text.lower()
    
    # Special cases/words
    text = re.sub(pattern="'til", repl="til", string=text)
    text = re.sub(pattern="'til", repl="til", string=text)
    text = re.sub(pattern="gon'", repl="gon", string=text)

    # Remove \n from beginning
    text = re.sub(pattern='^\n', repl='', string=text)

    # Strip , ! ?, : and remaining \n from lyrics
    text = ''.join([char.strip(",!?:") for char in text])
    text = text.replace('\n', ' ')

    # Remove contractions
    # specific
    text = re.sub(r"won't", "will not", text)
    text = re.sub(r"won\’t", "will not", text)
    text = re.sub(r"can't", "can not", text)
    text = re.sub(r"can\’t", "can not", text)
    text = re.sub(r"let's", "let us", text)
    text = re.sub(r"let\’s", "let us", text)
    text = re.sub(r"ain't", "aint", text)
    text = re.sub(r"ain\’t", "aint", text)

     # general
    text = re.sub(r"n\'t", " not", text)
    text = re.sub(r"\'re", " are", text)
    text = re.sub(r"\'s", " is", text)
    text = re.sub(r"\'d", " would", text)
    text = re.sub(r"\'ll", " will", text)
    text = re.sub(r"\'t", " not", text)
    text = re.sub(r"\'ve", " have", text)
    text = re.sub(r"\'m", " am", text)
    text = re.sub(r"n\’t", " not", text)
    text = re.sub(r"\’re", " are", text)
    text = re.sub(r"\’s", " is", text)
    text = re.sub(r"\’d", " would", text)
    text = re.sub(r"\’ll", " will", text)
    text = re.sub(r"\’t", " not", text)
    text = re.sub(r"\’ve", " have", text)
    text = re.sub(r"\’m", " am", text)

    # Remove remaining punctuation
    #punc = string.punctuation
   # text = ''.join([char for char in text if char not in punc])

    # Remove stopwords
    if remove_stops:
        stops = stopwords.words('english')
        text = ' '.join([word for word in text.split(' ') if word not in stops])
    
    # Remove double spaces and beginning/trailing whitespace
    text = re.sub(pattern='( ){2,}', repl=' ', string=text)
    text = text.strip()
    
    return(text)

In [ ]:
#Apply preprocessing function to data
df['clean_lyrics'] = df.apply(lambda x: preprocessText(x['lyrics']), axis=1)

In [ ]:
#Save processed dataframe to .csv
df.to_csv("lyrics_clean.csv")

In [ ]:
#Function for stemming the lyrics
df['stemmed_lyrics'] = df.apply(lambda x: stemming(x['lyrics']), axis=1)

Tokenization was not needed for the models below, as the CountVectorizer and TfidfVectorizer both tokenize the text on their own. Stemming was still performed as it improves model fit. 

# Classification Models (with CountVectorizer and TfidfVectorizer)

**Support Vector Machine (SVM)**

I trained an SVM model using SGDClassifier and TfidfVectorizer. The SVM model performed the second best of all models attempted, with slightly lower accuracy than the XGBoost model. 

In [ ]:
#SVM model with k-fold CV
text_svm = Pipeline([('vect', TfidfVectorizer(ngram_range=(1,2))),
                     ('svm', SGDClassifier(loss='hinge', penalty='l2', alpha=1e-4,
                                           max_iter=25, random_state=123))])
text_svm = text_svm.fit(train.stemmed_lyrics, train.genre)
cross_val_score(estimator=text_svm, X=train.stemmed_lyrics, y=train.genre, cv=7).mean()

0.6219237180115698

In [ ]:
#SVM test results
print(text_svm.score(y=test.genre, X=test.stemmed_lyrics))
preds_svm = text_svm.predict(test.stemmed_lyrics)
print(classification_report(y_pred=preds_svm, y_true=test.genre))
pd.crosstab(preds_svm, test.genre)

0.6215041464112097
              precision    recall  f1-score   support

     Country       0.58      0.86      0.69      3702
     Hip-Hop       0.73      0.86      0.79      3195
       Metal       0.64      0.84      0.73      3404
         Pop       0.56      0.42      0.48      3621
        Rock       0.54      0.16      0.25      3563

    accuracy                           0.62     17485
   macro avg       0.61      0.63      0.59     17485
weighted avg       0.61      0.62      0.58     17485



genre    Country  Hip-Hop  Metal   Pop  Rock
row_0                                       
Country     3179      105    134   941  1151
Hip-Hop       67     2746    151   467   313
Metal        177       92   2855   415   920
Pop          197      219    165  1521   613
Rock          82       33     99   277   566

The SVM model heavily misclassified Country as Rock and Pop, similarly to the Naive Bayes model. The lyrics in these genres can tend to be similar, so it is not too surprising. 

In [ ]:
#Which songs did the SVM model misclassify?
test['preds_svm'] = preds_svm
misclasses_svm = test[test.preds_svm != test.genre]
misclasses_svm['misclass_combo'] = misclasses_svm.apply(lambda x: x['genre']+'-'+x['preds_svm'], axis=1)

In [ ]:
#Number of misclasses per pair of genres. 
misclasses_svm.misclass_combo.value_counts()

Rock-Country       1151
Pop-Country         941
Rock-Metal          920
Rock-Pop            613
Pop-Hip-Hop         467
Pop-Metal           415
Rock-Hip-Hop        313
Pop-Rock            277
Hip-Hop-Pop         219
Country-Pop         197
Country-Metal       177
Metal-Pop           165
Metal-Hip-Hop       151
Metal-Country       134
Hip-Hop-Country     105
Metal-Rock           99
Hip-Hop-Metal        92
Country-Rock         82
Country-Hip-Hop      67
Hip-Hop-Rock         33
Name: misclass_combo, dtype: int64

**XGBoost**

Now I fit an XGBoost classifier model, in conjunction with the TfidfVectorizer. This model performed the best of all the models/parameterizations I attempted, with an accuracy of 63.3%.

In [ ]:
#XGB model
vect = TfidfVectorizer(ngram_range=(1,2))
vect.fit_transform(train.stemmed_lyrics)
vect_train = vect.transform(pd.Series(train.stemmed_lyrics))
vect_test = vect.transform(pd.Series(test.stemmed_lyrics))

In [ ]:
#Define XGBoost model
xgb = XGBClassifier(learning_rate=0.25, subsample=0.8, gamma=1, random_state=123, max_depth=4, max_delta_step=1).fit(vect_train, train.genre)

In [ ]:
#Look at results of XGBoost model
print(xgb.score(y=test.genre, X=vect_test))
preds_xgb = xgb.predict(vect_test)
print(classification_report(y_pred=preds_xgb, y_true=test.genre))
pd.crosstab(xgb.predict(vect_test), test.genre)

0.6327137546468401
              precision    recall  f1-score   support

     Country       0.64      0.73      0.68      3702
     Hip-Hop       0.88      0.79      0.83      3195
       Metal       0.70      0.73      0.72      3404
         Pop       0.54      0.51      0.52      3621
        Rock       0.43      0.42      0.43      3563

    accuracy                           0.63     17485
   macro avg       0.64      0.64      0.64     17485
weighted avg       0.63      0.63      0.63     17485



genre    Country  Hip-Hop  Metal   Pop  Rock
row_0                                       
Country     2699       70    116   547   753
Hip-Hop       16     2527     65   203    60
Metal        129       85   2495   235   599
Pop          371      344    220  1839   648
Rock         487      169    508   797  1503

The XGBoost model performed the best of all the models I attempted. It classified songs in the Hip-Hop (f1=0.83) and Metal (f1=0.72) genres the most accurately, while poorly classifying songs in the Rock genre (f1=0.43).

In [ ]:
#Which songs did the xgb model misclassify?
test['preds_xgb'] = preds_xgb
misclasses_xgb = test[test.preds_xgb != test.genre]
misclasses_xgb['misclass_combo'] = misclasses_xgb.apply(lambda x: x['genre']+'-'+x['preds_xgb'], axis=1)

In [ ]:
#Misclassified songs for each genre pair
misclasses_xgb.misclass_combo.value_counts()

Pop-Rock           797
Rock-Country       753
Rock-Pop           648
Rock-Metal         599
Pop-Country        547
Metal-Rock         508
Country-Rock       487
Country-Pop        371
Hip-Hop-Pop        344
Pop-Metal          235
Metal-Pop          220
Pop-Hip-Hop        203
Hip-Hop-Rock       169
Country-Metal      129
Metal-Country      116
Hip-Hop-Metal       85
Hip-Hop-Country     70
Metal-Hip-Hop       65
Rock-Hip-Hop        60
Country-Hip-Hop     16
Name: misclass_combo, dtype: int64

# Classification Using Word2Vec

**Word2Vec - Feature Extraction**

In [ ]:
#Import Dependencies 
import gensim
import pandas as pd
import nltk
import numpy as np
from nltk.corpus import brown
nltk.download('brown')

[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Unzipping corpora/brown.zip.


True

In [ ]:
#Tokenize lyrics using simple_preprocessor
from gensim.utils import simple_preprocess
# Tokenize the text column to get the new column 'tokenized_text'
df['tokenized_lyrics'] = [simple_preprocess(line, deacc=True) for line in df['clean_lyrics']] 
print(df['tokenized_lyrics'].head(10))

0     [love, crawdads, wheels, old, mag, fact, joint...
2     [female, newsreporter, talking, also, hearing,...
3     [ala, derecha, dale, suave, cabra³n, dale, sua...
4     [uh, ah, uh, uhh, hahaha, check, flipmode, squ...
5     [verse, getting, hard, tell, months, weeks, st...
6     [living, phat, pockets, flat, wit, tha, gat, r...
7     [word, bond, got, goin, amon, throw, ya, hands...
8     [gorilla, fucking, coupe, finna, pull, zoo, ni...
10    [many, brothers, fell, victim, streets, rest, ...
11    [ohhh, ohhh, ohhhhhh, ohh, ohhh, ohhhhh, heard...
Name: tokenized_lyrics, dtype: object


In [ ]:
#Stem the tokenized lyrics via PorterStemmer
from gensim.parsing.porter import PorterStemmer
porter_stemmer = PorterStemmer()
df['stemmed_tokens'] = [[porter_stemmer.stem(word) for word in tokens] for tokens in df['tokenized_lyrics'] ]

In [ ]:
#Create list of tokenized lyrics
Lyrics=list(df['clean_lyrics'])

In [ ]:
#Create corpus from the list of lyrics 
corpus=(Lyrics)

In [ ]:
#Tokenize and Create Vocabulary List using the lyrics and the very large Brown Corpus of words
sentence_data=[] 
Tokens=[]
for i in range(len(corpus)):
    g=corpus[i].split()
    sentence_data.append(set(g))
    Tokens.append((g))
#Brown Corpus Data
sentence_brown = brown.sents()   
for i in range(len(sentence_brown)):
    sentence_data.append(set(sentence_brown[i]))

In [ ]:
#Fit lyrics data to gensim model
model = gensim.models.Word2Vec(sentence_data, size=50,window=5,min_count=5)

In [ ]:
#Create the vocab list from the fitted Word2Vec model
vocab=model.wv.vocab
vocab=(set(vocab))

In [ ]:
#Apply Word2Vec to all of the data
word2vecTokens=[] 
i=0;
#Run a loop over all the tokens
for g in Tokens: 
    vc=[] #to store temp vector for each token in an instance
    for s in g:
        if (s in vocab):
            vc.append(model.wv[s])
    word2vecTokens.append(vc) # appending all the vectors of an instance

In [ ]:
g=[]
for i in range (len(word2vecTokens)):
    g.append(np.sum(word2vecTokens[i], axis=0)/len(word2vecTokens[i]))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: RuntimeWarning: invalid value encountered in double_scalars
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
#Create the columns and append our embedded tokens
columns=[]
index=[]
for k in range(0,57096):
    index.append(k)
for i in range(1,51):
    columns.append("w2v_"+str(i))
df_ = pd.DataFrame(columns=columns)
df_ = df_.fillna(0)

In [ ]:
for x in range (len(g)):
    try:
        if(type(g[x])==float):
            g[x]=[0]*50
        g[x]=g[x].tolist()
    except:
        print(x)

In [ ]:
#Convert the Word2Vec features into a pandas dataframe
for x in range (len(g)):
    try:
        if(type(g[x])==float):
            g[x]=[0]*50
    except:
        print(x)
w2v_df=pd.DataFrame(g,columns=columns)
w2v_df.head()

w2v_1     w2v_2     w2v_3  ...    w2v_48    w2v_49    w2v_50
0  0.209785  0.181589  0.110289  ...  0.313540 -1.144182 -0.270229
1  0.147199  0.214286  0.371582  ...  0.459570 -1.035808 -0.219823
2 -0.023518 -0.641457 -1.102616  ...  1.100754 -0.190145  1.338918
3  0.419271  0.081302  0.228797  ...  0.475145 -1.341440 -0.276306
4 -0.489338  0.686841 -0.446464  ...  1.322752 -1.471923 -0.246995

[5 rows x 50 columns]

In [ ]:
#Save Word2Vec data to .csv file
w2v_df.to_csv("w2vData.csv",index=False) # Saving the dataframe to a file

**Gradient Boosting Classification Using Word2Vec Features**

In [ ]:
#Import dependencies
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [ ]:
#Create column for genres in Word2Vec dataframe
w2v_df['genre'] = df['genre']

In [ ]:
#Drop all NAs
w2v_df = w2v_df.dropna(how='any')

In [ ]:
#Create train/test datasets
train,test,train_y,test_y = train_test_split(w2v_df[w2v_df.columns.difference(['genre'])],w2v_df['genre'],train_size=0.67)

In [ ]:
#Fit gradient boosting classification model
gd = GradientBoostingClassifier(max_depth=20)
gd.fit(train,train_y)
pred = gd.predict(test)
print (accuracy_score(test_y,pred))

0.4504994224366379


The Gradient Boosting Classifier achieved an accuracy of 45%, which is far lower than achieved by the XGBoost and SVM model. This is likely due to the fact that song lyrics often don't have cohesive word structures like, say, a book would. 

Other models that I tried included: Naive Bayes, SVM, Random Forests, and XGBoost. However, these all has even worse accuracy than the Gradient Boosting model. 